component/loki/source/kubernetes/kubetail/kubetail.go

```diff
@@ -0,0 +1,144 @@
+	Client *kubernetes.Clientset
+
+	// Handler to send discovered logs to.
+	Handler loki.EntryHandler
+
+	// Positions interface so tailers can save/restore offsets in log files.
+	Positions positions.Positions
+}
+
+// A Manager manages a set of running Tailers.
+type Manager struct {
+	log log.Logger
+
+	mut   sync.Mutex
+	opts  *Options
+	tasks []*tailerTask
+
+	runner *runner.Runner[*tailerTask]
+}
+
+// NewManager returns a new Manager which manages a set of running tailers.
+// Options must not be modified after passing it to a Manager.
+func NewManager(l log.Logger, opts *Options) *Manager {
+	return &Manager{
+		log: l,
+		runner: runner.New(func(t *tailerTask) runner.Worker {
+			return newTailer(l, t)
+		}),
+	}
+}
+
+// SyncTargets synchronizes the set of running tailers to the set specified by
+// targets.
+func (m *Manager) SyncTargets(ctx context.Context, targets []*Target) error {
+	m.mut.Lock()
+	defer m.mut.Unlock()
+
+	// Convert targets into tasks to give to the runner.
+	tasks := make([]*tailerTask, 0, len(targets))
+	for _, target := range targets {
+		tasks = append(tasks, &tailerTask{
+			Options: m.opts,
+			Target:  target,
+		})
+	}
+
+	if err := m.runner.ApplyTasks(ctx, tasks); err != nil {
+		return err
+	}
+
+	// Delete positions for targets which have gone away.
+	newEntries := make(map[positions.Entry]struct{}, len(targets))
+	for _, target := range targets {
+		newEntries[entryForTarget(target)] = struct{}{}
+	}
+
+	for _, task := range m.tasks {
+		ent := entryForTarget(task.Target)
+
+		// The task from the last call to SyncTargets is no longer in newEntries;
+		// remove it from the positions file. We do this _after_ calling ApplyTasks
+		// to ensure that the old tailers have shut down, otherwise the tailer
+		// might write its position again during shutdown after we removed it.
+		if _, found := newEntries[ent]; !found {
+			level.Info(m.log).Log("msg", "removing entry from positions file", "path", ent.Path, "labels", ent.Labels)
+			m.opts.Positions.Remove(ent.Path, ent.Labels)
+		}
+	}
+
+	m.tasks = tasks
+	return nil
+}
+
+func entryForTarget(t *Target) positions.Entry {
+	// Prefix the path with "cursor-" so it doesn't get automatically deleted by
+	// positions.Positions, which otherwise will try to read the path as a file.
+	return positions.Entry{
+		Path:   "cursor-" + t.String(),
+		Labels: t.Labels().String(),
+	}
+}
+
+// UpdateOptions updates the Options shared with all Tailers. All Tailers will
+// be updated with the new set of Options. Options should not be modified after
+// passing to UpdateOptions.
+func (m *Manager) UpdateOptions(ctx context.Context, newOptions *Options) error {
+	m.mut.Lock()
+	defer m.mut.Unlock()
+
+	// Iterate through the previous set of tasks and create a new task with the
+	// new set of options.
+	tasks := make([]*tailerTask, 0, len(m.tasks))
+	for _, oldTask := range m.tasks {
+		tasks = append(tasks, &tailerTask{
+			Options: newOptions,
```

Comment:
Should we be updating the same task with new Options (basically a new Client)?

It sounds like a rare edge-case, but I'm thinking that would mean that a pod with the same `namespace/name:container` in another cluster could potentially reuse an older positions entry from another cluster with a different `SinceTime` and this gap might lead to losing logs.

Reply:
Hm, I'm not sure, having the tasks be immutable is pretty nice to reason about. 

I'm not sure I'm following the edge case; are you talking about if someone changes the component to watch an entirely different cluster?

Reply:
Yes, that's the case I was thinking. It does sound pretty rare, but that's what would happen, right?